### PR TITLE
Start dev for entity generation (react)

### PR DESCRIPTION
**Pull request by qmonmert**
PR to say that I begin the dev to generate entity for react.

### CODE REVIEW

**Comment by deepu105** [comment]
I started working on them as well, but its fine if you would like to do it. Use this PR for it, i'll commit what I have here so you can use them. Follow the same pattern as the user management screen

**Comment by qmonmert** [comment]
ok, user management was already my target ;)

**Comment by deepu105** [comment]
I have commited what I had, you can continue with them :) thanks


Thanks & Regards,
Deepu

**Comment by qmonmert** [comment]
Thanks @deepu105 I continue with that ;)

**Comment by jdubois** [comment]
- FYI @qmonmert is working at Ippon next to me :-)
- For the entity sub-generator, I find what we have with Angular is too complex: we should generate "white applications" that are supposed to be easily modified by end-user to suit their business needs. Here we provide a beautiful, feature-rich UI, but that also makes it hard to modify. So for React, if we could have something simpler: I'm sure it will be better for end-users, and that should also help us to finish this sooner

**Comment by deepu105** [comment]
Lets start with simple entity screens first without any fancy
datepicker/file upload stuff first

Sendil, I didn't get what you were talking about, so lets discuss that when
we meet next time

File: generators/client/templates/react/src/main/webapp/app/config/constants.ts [hunk 1]
```diff
@@ -4,10 +4,14 @@ const config = {
...
+export const SERVER_API_URL = '';
+
 export const messages = {
   DATA_CREATE_SUCCESS_ALERT: 'Data created successfully!',
   DATA_DELETE_SUCCESS_ALERT: 'Data deleted successfully!',
   DATA_UPDATE_ERROR_ALERT: 'Data deleted successfully!'
 };
 
 export const APP_DATE_FORMAT = 'DD/MM/YY HH:mm';
+export const APP_FORMAT_LOCAL_DATE = 'DD/MM/YYYY';
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/config/constants.ts]: keep names consistent use `APP_LOCAL_DATE_FORMAT`, `APP_LOCAL_DATETIME_FORMAT`

File: generators/client/templates/react/src/main/webapp/app/config/translation.ts [hunk 2]
```diff
@@ -1,20 +1,32 @@
 /* global: require*/
 import { TranslatorContext } from 'react-jhipster';
+import * as _ from 'lodash';
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/config/translation.ts]: import only what is needed like `import { merge } from 'lodash';`

File: generators/client/templates/react/src/main/webapp/app/config/translation.ts [hunk 3]
```diff
@@ -1,20 +1,32 @@
...
 
 import { setLocale } from '../reducers/locale';
 import Storage from '../shared/util/storage-util';
 
-const mergeTranslations = requireContext => requireContext.keys().reduce(
-  (merged, key) => ({ ...merged, ...requireContext(key) }),
-  {}
-);
+// const mergeTranslations = requireContext => requireContext.keys().reduce(
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/config/translation.ts]: why are these changed? it has nothing to do with entities and I dont see why this change is needed as it will produce some weird code, please revert back to original code

File: generators/client/templates/react/src/main/webapp/app/entities/index.tsx [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import * as React from 'react';
+import { Route, Switch } from 'react-router-dom';
+import { ModalRoute } from 'react-router-modal';
+
+/* jhipster-needle-add-route-index-1 - JHipster will add routes here */
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/entities/index.tsx]: give a proper name `jhipster-needle-add-route-import`

File: generators/client/templates/react/src/main/webapp/app/entities/index.tsx [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+const Routes = ({ match }) => (
+  <div>
+    <Switch>
+        /* jhipster-needle-add-route-index-2 - JHipster will routes here */
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/entities/index.tsx]: give a proper name `jhipster-needle-add-route-path`

File: generators/client/templates/react/src/main/webapp/app/reducers/index.ts [hunk 6]
```diff
@@ -8,6 +8,7 @@ import layout from './layout';
 import authentication from './authentication';
 import administration from './administration';
 import userManagement from './user-management';
+/* jhipster-needle-add-reducer-1 - JHipster will add reducer here */
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/reducers/index.ts]: give a proper name `jhipster-needle-add-reducer-import`

File: generators/client/templates/react/src/main/webapp/app/reducers/index.ts [hunk 7]
```diff
@@ -18,4 +19,5 @@ export default combineReducers({
   administration,
   userManagement,
   loadingBar
+  /* jhipster-needle-add-reducer-2 - JHipster will add reducer here */
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/reducers/index.ts]: give a proper name `jhipster-needle-add-reducer-combine`

File: generators/client/templates/react/src/main/webapp/app/routes.tsx [hunk 8]
```diff
@@ -15,6 +16,7 @@ const Routes = () => (
     <Route path="/logout" component={Logout} />
     <PrivateRoute path="/admin" component={Admin} />
     <PrivateRoute path="/account" component={Account} />
+    <Route path="**" component={Entities} />
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/routes.tsx]: path should be `/entity`

File: generators/client/templates/react/src/main/webapp/app/shared/util/date-utils.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import * as moment from 'moment';
+
+import { APP_FORMAT_DATETIME_LOCAL } from '../../config/constants';
+
+export const convertDateTimeFromServer = (date: any) => {
+    if (date) {
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/shared/util/date-utils.ts]: change to 
```
export const convertDateTimeFromServer = date => 
  date ? moment(date).format(APP_FORMAT_DATETIME_LOCAL) : null;
```

type any is implicit so no need to specify

File: generators/entity-client/files.js [hunk 10]
```diff
@@ -22,9 +22,11 @@ const constants = require('../generator-constants');
 /* Constants use throughout */
 const CLIENT_TEST_SRC_DIR = constants.CLIENT_TEST_SRC_DIR;
 const ANGULAR_DIR = constants.ANGULAR_DIR;
+const REACT_DIR = constants.ANGULAR_DIR;
 
 const CLIENT_NG1_TEMPLATES_DIR = 'angularjs';
 const CLIENT_NG2_TEMPLATES_DIR = 'angular';
+const CLIENT_NG3_TEMPLATES_DIR = 'react';
```
Comment (deepu105) [inline generators/entity-client/files.js]: CLIENT_REACT_TEMPLATES_DIR

File: generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-detail.tsx [hunk 11]
```diff
@@ -0,0 +1,137 @@
+<%#
```
Comment (deepu105) [inline generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-detail.tsx]: add licence header to all new template files added

File: generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-detail.tsx [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+import { getEntity } from './<%= entityFileName %>.reducer';
+import { APP_DATE_FORMAT, APP_FORMAT_LOCAL_DATE } from '../../config/constants';
+
+export interface I<%= entityReactName %>DetailProps {
+  getEntity: ICrudGetAction;
+  <%= entityInstance %>: any;
+  match: any;
+}
+
+export class <%= entityReactName %>Detail extends React.Component<I<%= entityReactName %>DetailProps, undefined> {
```
Comment (deepu105) [inline generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-detail.tsx]: If there is no state then no need to define the generic just do `export class <%= entityReactName %>Detail extends React.Component<I<%= entityReactName %>DetailProps> {`

File: generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-dialog.tsx [hunk 13]
```diff
@@ -0,0 +1,350 @@
+import * as React from 'react';
+import { connect } from 'react-redux';
+import { Modal, ModalHeader, ModalBody, ModalFooter, Button, Label } from 'reactstrap';
+import { AvForm, AvGroup, AvInput, AvFeedback } from 'availity-reactstrap-validation';
+import { Translate } from 'react-jhipster';
+import { FaBan, FaFloppyO } from 'react-icons/lib/fa';
+
```
Comment (deepu105) [inline generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-dialog.tsx]: remove blank line here

File: generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-dialog.tsx [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+import * as moment from 'moment';
+
+import { ICrudGetAction, ICrudPutAction } from '../../shared/model/redux-action.type';
+import { getEntity, updateEntity, createEntity } from './<%= entityFileName %>.reducer';
+<%_ if (enableTranslation) { _%>
+import { locales } from '../../config/translation';
```
Comment (deepu105) [inline generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-dialog.tsx]: locales is not required here

File: generators/client/templates/react/src/main/webapp/app/config/translation.ts (hunk 3 above)
Comment (qmonmert) [reply generators/client/templates/react/src/main/webapp/app/config/translation.ts]: I changed this because when I generated several entities, only one file myEntityX.json (with the traduction) was used, the others were ignored :(

Comment (deepu105) [reply generators/client/templates/react/src/main/webapp/app/config/translation.ts]: it should fetch all files and merge them, as its doing the same currently and it works, don't worry about this just take care of entities and we can look into this later

**Comment by qmonmert** [comment]
@deepu105 if you want test react, you can import https://github.com/jhipster/jdl-samples/blob/master/simple-online-shop.jh or https://github.com/jhipster/jdl-samples/blob/master/blog.jh
Maybe it will be good to merge quickly like that others developers can also contribute or we can wait last week (I will work for sure on this until next friday) ? cc @jdubois

**Comment by deepu105** [comment]
Ok, ill have a final review and then will test and merge soon

**Comment by deepu105** [comment]
@qmonmert I fixed some issues to make a basic entity work with current master. I didnt find any issue with entity translation, i'll check again later, but I guess this version might be good enough so once travis passes I'll merge this

File: generators/client/templates/react/src/main/webapp/app/config/translation.ts [hunk 15]
```diff
@@ -1,10 +1,11 @@
 /* global: require*/
 import { TranslatorContext, Storage } from 'react-jhipster';
+import { merge } from 'lodash';
 
 import { setLocale } from '../reducers/locale';
 
 const mergeTranslations = requireContext => requireContext.keys().reduce(
-  (merged, key) => ({ ...merged, ...requireContext(key) }),
+  (merged, key) => merge({ ...merged }, { ...requireContext(key) }),
```
Comment (deepu105) [inline generators/client/templates/react/src/main/webapp/app/config/translation.ts]: @qmonmert this will solve the i18n issue you encountered

Comment (sendilkumarn) [reply generators/client/templates/react/src/main/webapp/app/config/translation.ts]: Cant we do this with `Object.assign` do we need `lodash` for this 🤔

File: generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-delete-dialog.tsx [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+-%>
+import * as React from 'react';
+import { connect } from 'react-redux';
+import { Modal, ModalHeader, ModalBody, ModalFooter, Button } from 'reactstrap';
+import { Translate, ICrudGetAction, ICrudDeleteAction } from 'react-jhipster';
+import { FaBan, FaTrash } from 'react-icons/lib/fa';
+
+import { getEntity, deleteEntity } from './<%= entityFileName %>.reducer';
+
+export interface I<%= entityReactName %>DeleteDialogProps {
```
Comment (sendilkumarn) [inline generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-delete-dialog.tsx]: The names are really long

File: generators/client/templates/react/src/main/webapp/app/config/translation.ts (hunk 15 above)
Comment (deepu105) [reply generators/client/templates/react/src/main/webapp/app/config/translation.ts]: Object.assign is not recursive, its same as { ... } destructuring which is only first level, we need deep merge here

File: generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-delete-dialog.tsx (hunk 16 above)
Comment (deepu105) [reply generators/entity-client/templates/react/src/main/webapp/app/entities/_entity-delete-dialog.tsx]: ya not sure what to do as we cant strip the props, and how much ever we try there will be some case where name is too long, let me see what can be done

**Comment by deepu105** [comment]
@sendilkumarn I overwrote the nit change you did as it was giving too many merge conflicts when I pulled it dies to rebase I have already done, those are minor things we can do later as we need to fix the formats anyway

**Comment by sendilkumarn** [comment]
Well okay. Let me merge this 👍
